evaluations/ipfs.md

```diff
@@ -0,0 +1,41 @@
+An evaluation of the IPFS paper
+===============================
+
+2021-01-07 Dagger Team
+
+https://ipfs.io/ipfs/QmR7GSQM93Cx5eAg6a6yRzNde1FQv7uL6X1o4k7zrJa3LX/ipfs.draft3.pdf
+
+Goal of this evaluation is to find things to adopt or avoid while designing
+Dagger. It is not meant to be a criticism of IPFS.
+
+#### Pros:
+
++ IPFS is designed by simplifying, evolving, and connecting proven techniques
+  (§3)
++ Consists of a stack of separately described sub-protocols (§3)
++ Uses Coral DSHT to favor data that is nearby, reducing latency of lookup
+  (§2.1.2)
++ Uses proof-of-work in S/Kademlia to discourage Sybil attacks (§2.1.3)
++ Favors self-describing values such as multihash (§3.1) and multiaddr (§3.2.1)
++ BitSwap protocol for exchanging blocks supports multiple strategies (§3.4.2),
+  so it should be relatively easy to add a micropayment strategy.
++ Uses content addressing (§3.5)
++ The Merkle DAG is simple, yet allows constucting filesystems,
+  key-value stores, databases, messaging system, etc.. (§3.5)
+
+#### Cons:
+
+- Kademlia prefers long-lived nodes (§2.1.1), which is not ideal for mobile
+  enviroments (although it's unclear whether there are any better alternatives)
+- The default BitSwap strategy falls just short of introducing a currency with
+  micro payments, necessitating additional work for nodes to find blocks to 
+  barter with (§3.4)
+- Object pinning (§3.5.3) inevitably leads to centralized gateways to IPFS, such
+  as Infura and Pinata
+- There are no self-describing multiformats for encryption and signing (§3.5.5),
```

Comment:
There is, it is described in `3.5.5 Object-level Cryptography`, for both signed and encrypted data.

Reply:
In 3.5.5 they describe a format for encryption and signatures, but I would not consider them to be self-describing. You can’t figure out which encryption/signature algorithm to use just by looking at an `EncryptedObject` or `SignedObject`, which is what I’d expect from a self-describing format.

Reply:
Well, `SignedObject` certainly does include all the info required to identify the scheme and verify it with the public key pointed to by the `PublicKey []multihash` field. For `EncryptedObject` you are right that it's impossible to tell what crypto primitive was used, but it seems to be a deliberate omission, you probably don't want to leak too much info about the encryption scheme.

```go
type SignedObject struct {
  Object []bytes
  // raw object data signed
  Signature []bytes
  // hmac signature
  PublicKey []multihash
  // multihash identifying key
}
```

Reply:
I still don't quite understand how the multihash for the public key tells you the signature scheme, but fair enough, there is much more info there than I initially thought. I'll remove the point, because it's now more of a nit-picking comment than an actual downside of using IPFS.